### ISSUE TITLE
Async rules and chaining

### ISSUE THREAD

**Issue by ajj7060** [MarimerLLC/cslaforum#663]
**Question**
How do async rules and chaining work together?

The specific scenario I have is that there's an async validation rule which should only run if the object is in a certain state.  Currently, the outer rule checks this state, and it is a sync rule.  The validation rule however is async.

Does this present any issues, or will it work as I expect it?

**Version and Platform**
CSLA version: 4.7.200
OS: Windows, iOS, Android
Platform: WinForms, WPF, Xamarin

**Comment by jonnybee** [comment]
It will not work as expected if you use RuleChaining (ie the async rule is an inner rule) of a sync rule as the rule engine will know about the sync rule. 

I would rather use RulePriority and have the async rule being the last rule to execute (f.ex with Priority=100) and create a ShortCircuiting rule at Priority 99 that short circuits rule processing when the object is not in the correct state. 

Example ShortCircuiting rule that will prevent rules with higher priority to run if the user is not allowed to edit this property. : 

    public class StopIfNotCanWrite : BusinessRule
    {
        public StopIfNotCanWrite(IPropertyInfo primaryProperty) : base(primaryProperty)
        { }  

        protected override void Execute(IRuleContext context)
        {
	    var target = (Csla.Core.BusinessBase)context.Target;

  	    // Short circuit rule processing for this property if user is not allowed to edit field.
	    if (!target.CanWriteProperty(PrimaryProperty))
	    {
	        context.AddSuccessResult(true);
	    }
        }
    }

**Comment by ajj7060** [comment]
Hi Jonny, thanks for confirming that I can't mix sync/async rules when chaining.  

Short circuiting doesn't work in my use case, as the outer sync rule is not a validation rule, it just determines if the inner rule needs to be run at all.  I'll have to combine the two rules into one I think and make it async.

**Comment by jonnybee** [comment]
ShortCircuiting works just fine. You control Priority so You only need to make sure the asynchronous rule runs after the short circuiting. I have used this technique often.
Once ShortCircuited the rule engine will stop all rocessing of rules for this property.

**Comment by jonnybee** [comment]
@ajj7060 The sync rule should just call context.AddSuccessResult(true); like i showed in the sample code. 

    public void AddSuccessResult(bool stopProcessing)
    {
      Results.Add(new RuleResult(Rule.RuleName, Rule.PrimaryProperty) 
            { Severity = RuleSeverity.Success, StopProcessing = stopProcessing });
    }

Assume that you add the async rule with Priority = 100 and the ShortCircuiting rule of 99 (so that these are always the last to run) then the only effect of the shortcircuiting rule is to control whether the rule engine will run the async rule or not.

**Comment by ajj7060** [comment]
@jonnybee Ah, the StopProcessing flag is what I was missing.  Does priority even matter for async rules though?  I thought they were always run last after all the sync rules anyway.

**Comment by jonnybee** [comment]
@ajj7060 Rules are always called in the order of Priority Ascending - doesn't matter if it is a sync or async rule.

**Comment by ajj7060** [comment]
@rockfordlhotka @jonnybee Has this always been the case, or has something changed?  Because this is what the Csla 4 book says "Synchronous rules are invoked according to priority order and are subject to short-circuiting. Asynchronous rules *are started after all synchronous rules for a property have been executed, and they run in an indeterminate order*."

**Comment by fujiiface** [comment]
I distinctly remember reading that in the book, as well.

**Comment by rockfordlhotka** [comment]
I'm pretty sure the book is correct.

**Comment by jonnybee** [comment]
This has always been the case, we are both right and nothing has changed. 

1. Priority is valid for both sync and async rules and the rule engine will always sort rules on a per-property (or object) level by Priority. The flag IsAsync is not used for sorting rules. 
2. An async rule may have some initial synchronous and at least one asynchronous call in its Execute method with a callback method to set the result i context and call context.Complete(). The initial synchrhronous code before the async call will be executed inline (sync) just like in a synchronous rule.  
3. It is **your** code in the BusinessRule.Execute that does some async operation. CSLA has no control over when the async call is executed - but the async callback will  **always** be called **after** all the synchronous rules have been executed. This is by design in .NET as the thread is busy and callback method cannot be called until the tread is free. This mashalling between threads is controlled by SynchronizationContext. 
4. If you look at the source code the main difference in handling is in the RuleContext.Complete() method.  When called from a sync rule it will update the OutputProperties and BrokenRules.  When called from an async rule it will updated the OutputProprties and BrokenRules and if it is the first run then also rerun rules for AffectedProperties and then also raise OnPropertyChanged for affected properties. 

A more accurate description is: 

> Business rules are invoked according to priority order. Only synchronous rules can cause short-circuiting. Asynchronous rules will always complete after all synchronous rules for a property have been executed and they run in an indeterminate order.

**Comment by ajj7060** [comment]
I did a sample, and Csla does honor the priority when starting rules, async or not. 

I think this should be considered a bug; starting the async rules in priority is really kinda meaningless when you don't know when they'll return, and presumably the rule is async to begin with because it takes a non trivial amount of time, so you'd want to save the expensive rules for last.  Plus I know we've been coding our rules with the assumption that the behavior matches what is documented in the bool.

**Comment by ajj7060** [comment]
@jonnybee I don't think that makes sense; you'd be better off just making all the rules sync.  The whole point of async is that they can run concurrently.
